ejb3/src/main/java/org/jboss/as/ejb3/remote/AssociationImpl.java

```diff
@@ -463,15 +464,34 @@ static Object invokeMethod(final ComponentView componentView, final Method metho
         final boolean isAsync = componentView.isAsynchronous(method);
         final boolean oneWay = isAsync && method.getReturnType() == void.class;
         final boolean isSessionBean = componentView.getComponent() instanceof SessionBeanComponent;
-        contextDataHolder.putAll(interceptorContext.getContextData());
+        final Object result;
         if (isAsync && isSessionBean) {
             if (! oneWay) {
                 interceptorContext.putPrivateData(CancellationFlag.class, cancellationFlag);
             }
-            final Object result = invokeWithIdentity(componentView, interceptorContext, securityIdentity);
+            result = invokeWithIdentity(componentView, interceptorContext, securityIdentity);
+            contextDataHolder.putAll(interceptorContext.getContextData());
+            sanitize(contextDataHolder);
             return result == null ? null : ((Future<?>) result).get();
         } else {
-            return invokeWithIdentity(componentView, interceptorContext, securityIdentity);
+            result = invokeWithIdentity(componentView, interceptorContext, securityIdentity);
+            contextDataHolder.putAll(interceptorContext.getContextData());
+            sanitize(contextDataHolder);
+            return result;
+        }
+    }
+
+    // Some objects added to the context data might cause harm if passed through remote calls.
+    // Obvious solution is to remove such objects in interceptors that added them.
+    // Current interceptor implementations, however, discourage that (see Jsr299BindingsCreateInterceptor) and require
+    // modification of many classes in a chain through which added data is passed to allow cleanup actions.
+    private static void sanitize(Map<String, Object> map) {
+        final Iterator<Map.Entry<String, Object>> it = map.entrySet().iterator();
+        while (it.hasNext()) {
+            final Map.Entry<String, Object> entry = it.next();
+            if (entry.getValue().getClass().getName().contains("org.jboss.weld.context.SerializableContextualInstanceImpl")) {
```

Comment:
Yeah, it looks this ugly because key is dynamically generated based on interceptor's bean class.